### DISCUSSION TITLE
custom gate 阶数、对 proof 性能的影响

### DISCUSSION THREAD

**Discussion by mdj33** [z2o-k7e/halo2-step-by-step#17] [Q&A]
不同custom gate 需要query的advice column不同，但也可能会有重叠。
所有的custom gate的poly最终会累加起来，一般来说单纯累加对性能影响有限， 那么更多的custom gate会做更多的类似FFT的计算吗，会对性能产生很大影响吗

> shuidima:  "custom gate不宜阶数过大，现在一般最大是4阶，如果需要更复杂的表达，需要拆成多个express，但是从数学原理上还是比较模糊， 比如类似plonk里面的 qMab, 这里插值成多项式时候，就会有3n次的多项式，由于超出了FFT支持的n次，需要把3n次的多项式拆成3个n次的来算，如果custom gate到4n，5n，就要拆成4个n次，5个n次多项式来算，这样会增加FFT的计算时间，但是个人理解的比较模糊，没有真正从数学上理解清楚，开会时候可以请教下老师"

**Comment by Demian101** [comment]
@Cecilia：
https://github.com/privacy-scaling-explorations/halo2/blob/9eaccbb8578f3c1741d4961eef7405a4e80a9dfe/halo2_proofs/src/poly/domain.rs#L62

正好在搞benchmark 看来大家问题问的很精准[ThumbsUp]

@Dream:


不是 直接在3n(4n, 5n)的extend domain上来做fft
